File: apply-migrations.py
#!/usr/bin/env python3
import os
import sys
import psycopg2
import re

# Load .env file
env = {}
if os.path.exists('.env'):
    with open('.env', 'r') as f:
        for line in f:
            if line.strip() and not line.startswith('#'):
                key, *value_parts = line.split('=')
                env[key.strip()] = '='.join(value_parts).strip()

supabase_url = env.get('NEXT_PUBLIC_SUPABASE_URL')
service_role_key = env.get('SUPABASE_SERVICE_ROLE_KEY')

if not supabase_url or not service_role_key:
    print("❌ Missing SUPABASE environment variables in .env")
    sys.exit(1)

# Extract project ID and region from Supabase URL
# URL format: https://[project-ref].supabase.co
match = re.search(r'https://([a-z0-9]+)\.supabase\.co', supabase_url)
if not match:
    print("❌ Invalid Supabase URL format")
    sys.exit(1)

project_ref = match.group(1)

# Supabase PostgreSQL connection details
# Default port: 5432
# Database: postgres
# User: postgres
# Password is in the service role JWT, but we need the actual postgres password
# Actually, let me try using psql with the connection string

print(f"🔗 Connecting to Supabase PostgreSQL (project: {project_ref})")
print(f"   URL: {supabase_url}")

# Try to connect using Supabase's PostgreSQL endpoint
# For Supabase, the connection string is:
# postgresql://postgres:[password]@db.[project-ref].supabase.co:5432/postgres

# We don't have the postgres password, but we can try connecting without it
# or using the service role key's decoded info

try:
    from jwt import decode as jwt_decode

    # Decode the service role JWT to get potential info
    decoded = jwt_decode(service_role_key, options={"verify_signature": False})
    print(f"   Service role decoded: role={decoded.get('role')}")
except Exception as e:
    print(f"   Could not decode JWT: {e}")

print("\n⚠️  Direct PostgreSQL connection requires database password")
print("    which is not available in the .env file.")
print("\n📋 ALTERNATIVE: Please manually apply migrations via Supabase Dashboard\n")
print("   See: MIGRATION_INSTRUCTIONS.md\n")
print("   Steps:")
print("   1. Go to https://app.supabase.com")
print("   2. Select project: tdedjksugbqgkxsrzzjo")
print("   3. SQL Editor → New Query")
print("   4. Copy-paste each migration file and run\n")

File: apply-migrations.js
const { createClient } = require('@supabase/supabase-js');
const fs = require('fs');
const path = require('path');
require('dotenv').config();

const supabaseUrl = process.env.NEXT_PUBLIC_SUPABASE_URL;
const serviceRoleKey = process.env.SUPABASE_SERVICE_ROLE_KEY;

if (!supabaseUrl || !serviceRoleKey) {
  console.error('❌ Missing SUPABASE environment variables');
  process.exit(1);
}

console.log(`🔗 Connecting to Supabase: ${supabaseUrl}`);
const supabase = createClient(supabaseUrl, serviceRoleKey);

const migrationFiles = [
  'migrations/001_create_auth_tables.sql',
  'migrations/002_add_email_fields_to_wa_sender_sessions.sql',
  'migrations/003_add_sent_status_column.sql',
];

async function runMigrations() {
  try {
    console.log('🚀 Applying Supabase migrations...\n');

    for (const file of migrationFiles) {
      const filePath = path.join(__dirname, file);
      if (!fs.existsSync(filePath)) {
        console.error(`❌ Migration file not found: ${file}`);
        continue;
      }

      const sql = fs.readFileSync(filePath, 'utf8');
      console.log(`📝 Running: ${file}`);

      // Execute raw SQL via PostgreSQL connection
      const { error } = await supabase.rpc('exec', { sql }).catch(async () => {
        // Fallback: split by statements and execute individually
        const statements = sql
          .split(';')
          .map(s => s.trim())
          .filter(s => s.length > 0 && !s.startsWith('--'));

        for (const statement of statements) {
          const { error: stmtError } = await supabase
            .from('_sql_raw')
            .select('*')
            .single()
            .catch(() => ({ error: null }));

          // Try via admin API instead
          try {
            const response = await fetch(`${supabaseUrl}/rest/v1/rpc/execute_sql`, {
              method: 'POST',
              headers: {
                'Authorization': `Bearer ${serviceRoleKey}`,
                'Content-Type': 'application/json',
                'apikey': serviceRoleKey,
              },
              body: JSON.stringify({ sql: statement }),
            });

            if (!response.ok) {
              const responseText = await response.text();
              console.log(`   → Statement: ${statement.substring(0, 60)}...`);
              if (responseText.includes('already exists') || responseText.includes('If not exists')) {
                console.log(`   ✓ OK (already exists)`);
              }
            } else {
              console.log(`   ✓ Executed`);
            }
          } catch (e) {
            console.log(`   ⚠ Statement may have failed: ${e.message}`);
          }
        }
        return { error: null };
      });

      if (error) {
        console.log(`⚠️  Note: ${error.message}`);
      } else {
        console.log(`✅ Applied: ${file}\n`);
      }
    }

    // Verify tables exist
    console.log('\n🔍 Verifying schema...');
    const { data: tables, error: tableError } = await supabase
      .from('information_schema.tables')
      .select('table_name')
      .eq('table_schema', 'public')
      .in('table_name', ['users', 'wa_sender_sessions']);

    if (tableError) {
      console.log('⚠️  Could not verify tables via direct query');
    } else if (tables && tables.length >= 2) {
      console.log(`✅ Tables verified:`);
      tables.forEach(t => console.log(`   • ${t.table_name}`));
    } else {
      console.log('⚠️  Tables may not have been created');
    }

    console.log('\n✅ Migration process completed!');
    console.log('   If you see errors above about "already exists", that\'s OK - it means the schema is already in place.');
  } catch (error) {
    console.error('❌ Migration failed:', error.message);
    process.exit(1);
  }
}

runMigrations();
